src/CustomElementDefinitionProducer.js

```diff
@@ -0,0 +1,136 @@
+
+  /**
+   *
+   * @param {string} localName
+   * @param {!Function} constructorOrGetter
+   * @param {boolean} isGetter
+   * @param {!CustomElementInternals} internals
+   */
+  constructor(localName, constructorOrGetter, isGetter, internals) {
+
+    /**
+     * @private
+     * @const {!CustomElementInternals}
+     */
+    this._internals = internals;
+
+    /**
+     * @private
+     * @const {!Function}
+     */
+    this._constructorOrGetter = constructorOrGetter;
+
+    /**
+     * @private
+     * @const {boolean}
+     */
+    this._isGetter = isGetter;
+
+    /** @const {string} */
+    this.localName = localName;
+
+    /**
+     * @private
+     * @type {!CustomElementDefinition|undefined}
+     */
+    this._definition;
+
+    if (!(constructorOrGetter instanceof Function)) {
+      throw new TypeError('Custom element constructors must be functions.');
+    }
+
+    if (!Utilities.isValidCustomElementName(localName)) {
+      throw new SyntaxError(`The element name '${localName}' is not valid.`);
+    }
+
+    if (this._internals.localNameToDefinitionProducer(localName)) {
+      throw new Error(`A custom element with name '${localName}' has already been defined.`);
+    }
+
+    if (elementDefinitionIsRunning) {
+      throw new Error('A custom element is already being defined.');
+    }
+
+    // Per spec, reify the definition immediately if possible.
+    if (!this._isGetter) {
+      this._ensureDefinition();
+    }
+  }
+
+  /** @return {!CustomElementDefinition} */
+  get definition() {
+    this._ensureDefinition();
+    return /** @type {!CustomElementDefinition} */ (this._definition);
+  }
+
+  _ensureDefinition() {
+    if (this._definition) {
+      return;
+    }
+    elementDefinitionIsRunning = true;
+    let connectedCallback;
+    let disconnectedCallback;
+    let adoptedCallback;
+    let attributeChangedCallback;
+    let observedAttributes;
+    let constructorFunction;
+    let error;
+    try {
+      constructorFunction = this._isGetter ? this._constructorOrGetter() :
+        this._constructorOrGetter;
+      /** @type {!Object} */
+      const prototype = constructorFunction.prototype;
+      if (!(prototype instanceof Object)) {
+        throw new TypeError('The custom element constructor\'s prototype is not an object.');
+      }
+
+      function getCallback(name) {
+        const callbackValue = prototype[name];
+        if (callbackValue !== undefined && !(callbackValue instanceof Function)) {
+          throw new Error(`The '${name}' callback must be a function.`);
+        }
+        return callbackValue;
+      }
+      connectedCallback = getCallback('connectedCallback');
+      disconnectedCallback = getCallback('disconnectedCallback');
+      adoptedCallback = getCallback('adoptedCallback');
+      attributeChangedCallback = getCallback('attributeChangedCallback');
+      observedAttributes = constructorFunction['observedAttributes'] || [];
+    } catch (e) {
```

Comment:
`define` is specced to re-throw exceptions that are thrown from inside any user code it calls<sup>1</sup>, in which case no definition is added to the registry's map.

So, you can do something like this:

```javascript
document.body.appendChild(document.createElement('custom-element'));

try {
  customElements.define('custom-element', class extends HTMLElement {
    constructor() {
      super();
      console.log("First definition constructed.");
    }
    static get observedAttributes() {
      throw new Error('Some error...');
    }
    // `observedAttributes` is only read if `attributeChangedCallback` exists. ^2
    attributeChangedCallback() {}
  });
} catch (e) {
  console.log('oops: ', e);
}

customElements.define('custom-element', class extends HTMLElement {
  constructor() {
    super();
    console.log("Second definition constructed.");
  }
});
```

... and the `<custom-element>` added in the first line will be upgraded with the second definition.

I think this means that this function should still bail when this catch happens and we'll need to update `get definition` to return `{!CustomElementDefinition|undefined}` as well as things that depend on it. Also, `this._internals.localNameToDefinitionProducer(localName)` doesn't completely determine if a definition already exists for an element anymore: you could have an existing producer for a lazy definition that will fail but hasn't been run yet. It would need to both check if the producer exists as well as force the producer to get the definition:

```javascript
const existingProducer = this._internals.localNameToDefinitionProducer(localName);
if (existingProducer && producer.definition) {
  throw new Error(`A custom element with name '${localName}' has already been defined.`);
}
```

We should pull the producer out of the map in CustomElementInternals when it fails in this way.

---

<sup>1</sup> Step 14 of [`define`](https://html.spec.whatwg.org/multipage/custom-elements.html#dom-customelementregistry-define) is run while catching any exceptions, then throws any caught exceptions as the last sub-step. The polyfill is incorrect and it suppresses these errors. This could probably be fixed by changing this `return;` to `throw e;`:
https://github.com/webcomponents/custom-elements/blob/92f66f713dc29e0428855ed7eaf4b1e5e02a3e92/src/CustomElementRegistry.js#L118

<sup>2</sup> The polyfill is incorrect and always reads `observedAttributes`:
https://github.com/webcomponents/custom-elements/blob/92f66f713dc29e0428855ed7eaf4b1e5e02a3e92/src/CustomElementRegistry.js#L116